src/frontend/src/optimizer/plan_node/stream_locality_provider.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::Itertools;
+use risingwave_common::catalog::Field;
+use risingwave_common::hash::VirtualNode;
+use risingwave_common::types::DataType;
+use risingwave_common::util::sort_util::OrderType;
+use risingwave_pb::stream_plan::LocalityProviderNode;
+use risingwave_pb::stream_plan::stream_node::PbNodeBody;
+
+use super::stream::prelude::*;
+use super::utils::{TableCatalogBuilder, impl_distill_by_unit};
+use super::{ExprRewritable, PlanTreeNodeUnary, StreamNode, StreamPlanRef as PlanRef, generic};
+use crate::TableCatalog;
+use crate::catalog::TableId;
+use crate::expr::{ExprRewriter, ExprVisitor};
+use crate::optimizer::plan_node::PlanBase;
+use crate::optimizer::plan_node::expr_visitable::ExprVisitable;
+use crate::optimizer::property::Distribution;
+use crate::stream_fragmenter::BuildFragmentGraphState;
+
+/// `StreamLocalityProvider` implements [`super::LogicalLocalityProvider`]
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct StreamLocalityProvider {
+    pub base: PlanBase<Stream>,
+    core: generic::LocalityProvider<PlanRef>,
+}
+
+impl StreamLocalityProvider {
+    pub fn new(core: generic::LocalityProvider<PlanRef>) -> Self {
+        let input = core.input.clone();
+
+        let dist = match input.distribution() {
+            Distribution::HashShard(keys) => {
+                // If the input is hash-distributed, we make it a UpstreamHashShard distribution
+                // just like a normal table scan. It is used to ensure locality provider is in its own fragment.
```

Comment:
Does it mean that this is actually a workaround for the backfill order control to work correctly? I guess we can keep the `HashShard` and omit one shuffle if it's solely for correctness.